resolver.js

```diff
@@ -5,62 +5,41 @@ const { getAllUsers, getUserByNameAndPassword, addUser, getUser } = require('./s
 
 module.exports = {
   Query: {
-    fetchUserCoupon(_, args) {
-      const { userId } = args;
-      const couponsToReturn = [];
+    async fetchCoupons(_, { userId }) {
+      let coupons = [];
 
       if (!userId) {
-        couponsToReturn.push(...getAllCoupons());
+        coupons = await getAllCoupons();
       } else {
-        couponsToReturn.push(...getUserCoupons(userId));
+        coupons = await getUserCoupons(userId);
       }
 
-      return couponsToReturn;
+      return coupons;
     },
-    fetchCoupon(_, args) {
-      const { id } = args;
-
-      const coupon = getCoupon(id);
+    async fetchCoupon(_, { id }) {
+      const coupon = await getCoupon(id);
 
       if (coupon) return coupon;
       throw new Error('Coupon not found');
     },
-    fetchNonSecureUsers: () => getAllUsers(),
+    fetchNonSecureUsers: async () => await getAllUsers(),
   },
   Mutation: {
-    signin(_, args) {
-      const { name, password } = args;
-      const user = getUserByNameAndPassword(name, password);
+    async signin(_, { name, password }) {
+      const user = await getUserByNameAndPassword(name, password);
 
       if (user) return user;
       throw new Error('User not found.')
     },
-    signup(_, args) {
-      const { name, password } = args;
-      const user = addUser(name, password);
-
-      return user;
-    },
-    addCoupon(_, args) {
-      const { userId, couponCode, expiry } = args;
-      const coupon = addCoupon(couponCode, expiry, userId);
-
-      return coupon;
+    async deleteCoupon(_, { id, userId }) {
+      await deleteCoupon(id, userId);
+      return true;
     },
-    editCoupon(_, args) {
-      const { id, userId, couponCode, expiry } = args;
-      const coupon = editCoupon(id, couponCode, expiry, userId);
-
-      return coupon;
-    },
-    deleteCoupon(_, args) {
-      const { id, userId } = args;
-      const coupon = deleteCoupon(id, userId);
-
-      return coupon;
-    }
+    signup: async (_, { name, password }) => await addUser(name, password),
```

Comment:
You don't need to await the content of your explicit return.
So you can rewrite this as 

```javascript
signup: async (_, { name, password }) => addUser(name, password)
```

There's not really a bad effect on doing this but I know I read somewhere how this can cause a small lag so can't remember, just thought to mention it.

Reply:
All that matters is that the resolver is async, I think JS will automatically await the return value here.

Reply:
Noted.

Is it a `JavaScript` feature though? Or a `GraphQL` feature?
Because I don't think JS has **"implicit await"** as a feature. I might be wrong though.

Reply:
So...

Apologies I didn't explain it well. If you have an async method that returns an async method. When you return it you don't need to call `await` because wherever the outer function is being called, once it's awaited the return function will also be run and until it's returned before it completes execution.

I try to use code to show [the thing here](https://repl.it/join/xtxovyad-cooproton).